common/onion.c

```diff
@@ -0,0 +1,304 @@
+	}
+}
+
+/* Returns true if valid, and fills in type. */
+static bool pull_payload_length(const u8 **cursor,
+				size_t *max,
+				enum onion_payload_type *type,
+				size_t *len)
+{
+	/* *len will incorporate bytes we read from cursor */
+	const u8 *start = *cursor;
+
+	/* BOLT #4:
+	 *
+	 * The `length` field determines both the length and the format of the
+	 * `hop_payload` field; the following formats are defined:
+	 */
+	*len = fromwire_bigsize(cursor, max);
+	if (!cursor)
+		return false;
+
+	/* BOLT #4:
+	 * - Legacy `hop_data` format, identified by a single `0x00` byte for
+	 *   length. In this case the `hop_payload_length` is defined to be 32
+	 *   bytes.
+	 */
+	if (*len == 0) {
+		if (type)
+			*type = ONION_V0_PAYLOAD;
+		assert(*cursor - start == 1);
+		*len = 1 + 32;
+		return true;
+	}
+
+#if !EXPERIMENTAL_FEATURES
+	/* Only handle legacy format */
+	return false;
+#else
+	/* BOLT #4:
+	 * - `tlv_payload` format, identified by any length over `1`. In this
+	 *   case the `hop_payload_length` is equal to the numeric value of
+	 *   `length`.
+	 */
+	if (*len > 1) {
+		/* It's still invalid if it claims to be too long! */
+		if (*len > ROUTING_INFO_SIZE - HMAC_SIZE)
+			return false;
+
+		if (type)
+			*type = ONION_TLV_PAYLOAD;
+		*len += (*cursor - start);
+		return true;
+	}
+
+	return false;
+#endif /* EXPERIMENTAL_FEATURES */
+}
+
+size_t onion_payload_length(const u8 *raw_payload, size_t len,
+			    bool *valid,
+			    enum onion_payload_type *type)
+{
+	size_t max = len, payload_len;
+	*valid = pull_payload_length(&raw_payload, &max, type, &payload_len);
+
+	/* If it's not valid, copy the entire thing. */
+	if (!*valid)
+		return len;
+
+	return payload_len;
+}
+
+struct onion_payload *onion_decode(const tal_t *ctx,
+				    const struct route_step *rs)
+{
+	struct onion_payload *p = tal(ctx, struct onion_payload);
+	const u8 *cursor = rs->raw_payload;
+	size_t max = tal_bytelen(cursor), len;
+	struct tlv_tlv_payload *tlv;
+
+	if (!pull_payload_length(&cursor, &max, &p->type, &len))
+		return tal_free(p);
+
+	switch (p->type) {
+	case ONION_V0_PAYLOAD:
+		p->type = ONION_V0_PAYLOAD;
+		p->forward_channel = tal(p, struct short_channel_id);
+		fromwire_short_channel_id(&cursor, &len, p->forward_channel);
+		p->amt_to_forward = fromwire_amount_msat(&cursor, &len);
+		p->outgoing_cltv = fromwire_u32(&cursor, &len);
+		p->payment_secret = NULL;
+
+		if (rs->nextcase == ONION_FORWARD) {
+			p->total_msat = NULL;
+		} else {
+			/* BOLT-e36f7b6517e1173dcbd49da3b516cfe1f48ae556 #4:
+			 * - if it is the final node:
+			 *   - MUST treat `total_msat` as if it were equal to
+			 *     `amt_to_forward` if it is not present. */
+			p->total_msat = tal_dup(p, struct amount_msat,
+						&p->amt_to_forward);
+		}
+		return p;
+
+	case ONION_TLV_PAYLOAD:
+		tlv = tal(p, struct tlv_tlv_payload);
+		if (!fromwire_tlv_payload(&cursor, &len, tlv))
+			return tal_free(p);
+		if (!tlv_payload_is_valid(tlv, NULL))
+			return tal_free(p);
```

Comment:
This needs to go, it's still pre-hook call, and too early to apply our checks. These should later done in `htlc_accepted_hook_callback`, otherwise we never call the hook.